Mathlib/NumberTheory/Zagier.lean

```diff
@@ -0,0 +1,278 @@
+    dsimp
+  split_ifs with less more <;> (unfold complexInvo; simp; congr)
+  · simp [show ¬(x + 2 * z + (y - x - z) < z) by linarith [less], xb]
+    rw [Nat.sub_sub, two_mul, ← tsub_add_eq_add_tsub (by linarith), ← add_assoc,
+      Nat.add_sub_cancel, add_comm (x + z), Nat.sub_add_cancel (less.le)]
+  · push_neg at less
+    simp [(show x - 2 * y + y < x + z - y by zify [more.le, less]; linarith)]
+    constructor
+    · rw [Nat.sub_add_cancel more.le]
+    · rw [Nat.sub_right_comm, Nat.sub_sub _ _ y, ← two_mul, add_comm, Nat.add_sub_assoc more.le,
+        Nat.add_sub_cancel]
+  · push_neg at less more
+    simp [(show ¬(2 * y - x + (x + z - y) < y) by push_neg; zify [less, more]; linarith),
+      (show ¬(2 * y < 2 * y - x) by push_neg; zify [more]; linarith)]
+    constructor
+    · rw [tsub_tsub_assoc (by rfl) more, tsub_self, zero_add]
+    · rw [← Nat.add_sub_assoc less (2 * y - x), ← add_assoc, Nat.sub_add_cancel more,
+        Nat.sub_sub _ _ y, ← two_mul, add_comm, Nat.add_sub_cancel]
+
+theorem unique {t : zagierSet k}
+    (mem : t ∈ (MulAction.fixedPoints (powers (complexInvo k)) (zagierSet k))) :
+    t.val = (1, 1, k):= by
+  simp only [MulAction.mem_fixedPoints, Subtype.forall] at mem
+  replace mem := mem (complexInvo k) (mem_powers _)
+  rw [Submonoid.smul_def, End.smul_def] at mem
+  unfold complexInvo at mem
+  obtain ⟨⟨x, y, z⟩, h⟩ := t
+  obtain ⟨xb, yb, zb⟩ := zagierSet_lower_bound k h
+  apply_fun Subtype.val at mem
+  simp at mem
+  split_ifs at mem with less more <;> simp_all
+  · obtain ⟨_, _, _⟩ := mem
+    simp_all
+  · -- True case
+    unfold zagierSet at h; simp at h
+    replace mem := mem.1
+    rw [tsub_eq_iff_eq_add_of_le more, ← two_mul] at mem
+    replace mem := (mul_left_cancel₀ two_ne_zero mem).symm
+    subst mem
+    rw [show x * x + 4 * x * z = x * (x + 4 * z) by linarith] at h
+    cases' (Nat.dvd_prime hk.out).1 (dvd_of_mul_left_eq _ h) with e e
+    · rw [e, mul_one] at h; subst h
+      simp_all [show z = 0 by linarith [e]]
+    · rw [e, mul_left_eq_self₀] at h; simp_all
+      linarith [e]
+
+theorem aux :
+    MulAction.fixedPoints (powers (complexInvo k)) (zagierSet k) =
+    {⟨(1, 1, k), (by unfold zagierSet; simp; linarith)⟩} := by
+  rw [Set.eq_singleton_iff_unique_mem]
+  constructor
+  · simp
+    intro f h
+    rw [Submonoid.smul_def, End.smul_def]
+    simp
+    rw [mem_powers_iff] at h
+    obtain ⟨n, h⟩ := h
+    cases' Nat.even_or_odd n with heven hodd <;> rw [pow_eq_iterate] at h
+    · rw [(involutive_complexInvo k).iterate_even heven] at h
+      rw [← h, id_eq]
+    · rw [(involutive_complexInvo k).iterate_odd hodd] at h
+      rw [← h]
+      unfold complexInvo
+      simp
+  · intro t mem
+    replace mem := unique k mem
+    unfold zagierSet at t
+    congr!
+
+theorem positive_of_odd {n : ℕ} (h : n % 2 = 1) : 0 < n := by
+  by_contra j; push_neg at j; rw [nonpos_iff_eq_zero] at j; subst j; norm_num at h
+
+theorem result : ∃ a b : ℕ, a ^ 2 + b ^ 2 = 4 * k + 1 := by
```

Comment:
There is no namespacing, so this result will just be in the global namespace called `result`. Please give it a descriptive name and namespace it also (and maybe also make it `protected` if its name overlaps too much).
Similar with other results